flocker/acceptance/endtoend/test_leases.py

```diff
@@ -0,0 +1,229 @@
+# Copyright ClusterHQ Inc.  See LICENSE file for details.
+
+"""
+Tests for the leases API.
+"""
+
+from uuid import UUID, uuid4
+
+from twisted.internet import reactor
+from twisted.internet.task import deferLater
+from twisted.trial.unittest import TestCase
+
+from docker.utils import create_host_config
+
+from ...testtools import random_name
+from ..testtools import (
+    require_cluster, require_moving_backend, create_dataset,
+    REALISTIC_BLOCKDEVICE_SIZE, get_docker_client,
+    post_http_server, assert_http_server,
+)
+from ..scripts import SCRIPTS
+
+
+class LeaseAPITests(TestCase):
+    """
+    Tests for the leases API.
+    """
+    timeout = 600
+
+    def _assert_lease_behavior(self, cluster, operation,
+                               additional_kwargs, state_method):
+        """
+        Assert that leases prevent datasets from being moved or deleted.
+
+        * Create a dataset on node1.
+        * Acquire a lease for dataset on node1.
+        * Start a container (directly using docker-py) bind mounted to dataset
+          mount point on node1 and verify that data can be written.
+        * Stop the container.
+        * Request a move or delete operation.
+        * Wait for a short time; enough time for an unexpected unmount to take
+          place.
+        * Restart the container and write data to it, to demonstrate that the
+          dataset is still mounted and writable.
+        * Stop the container again.
+        * Release the lease, allowing the previously requested operation to
+          proceed.
+        * Wait for the previously requested operation to complete.
+
+        :param Cluster cluster: The cluster on which to operate.
+        :param operation: The ``FlockerClient`` method to call before releasing
+            the lease.
+        :param dict additional_kwargs: Any additional arguments to pass to
+            ``operation``.
+        :param state_method: A callable which returns a ``Deferred`` that fires
+            when the requested operation has been performed.
+        :returns: A ``Deferred`` that fires when the all the steps have
+            completed.
+        """
+        http_port = 8080
```

Comment:
What happens if something is already listening on this port? 

JP recently added `add_with_port_collision_retry` to `flocker.node.testtools`. Could we use that here to make the test more reliable?